arangod/Aql/OutputAqlItemRow.h

```diff
@@ -0,0 +1,156 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2018 ArangoDB GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Tobias Gödderz
+/// @author Michael Hackstein
+/// @author Heiko Kernbach
+/// @author Jan Christoph Uhde
+////////////////////////////////////////////////////////////////////////////////
+
+#ifndef ARANGOD_AQL_OUTPUT_AQL_ITEM_ROW_H
+#define ARANGOD_AQL_OUTPUT_AQL_ITEM_ROW_H
+
+#include "Aql/AqlItemBlock.h"
+#include "Aql/ExecutorInfos.h"
+#include "Aql/InputAqlItemRow.h"
+#include "Aql/types.h"
+#include "Basics/Common.h"
+
+namespace arangodb {
+namespace aql {
+
+struct AqlValue;
+
+/**
+ * @brief One row within an AqlItemBlock, for writing.
+ *
+ * Does not keep a reference to the data.
+ * Caller needs to make sure that the underlying
+ * AqlItemBlock is not going out of scope.
+ */
+class OutputAqlItemRow {
+ public:
+  explicit OutputAqlItemRow(
+      std::unique_ptr<OutputAqlItemBlockShell> blockShell);
+
+  void setValue(RegisterId registerId, InputAqlItemRow const& sourceRow,
+                AqlValue const&);
+
+  void copyRow(InputAqlItemRow const& sourceRow);
+
+  std::size_t getNrRegisters() const { return block().getNrRegs(); }
+
+  /**
+   * @brief May only be called after all output values in the current row have
+   * been set, or in case there are zero output registers, after copyRow has
+   * been called.
+   */
+  void advanceRow() {
+    TRI_ASSERT(produced());
+    if (!allValuesWritten()) {
+      THROW_ARANGO_EXCEPTION(TRI_ERROR_WROTE_TOO_FEW_OUTPUT_REGISTERS);
+    }
+    if (!_inputRowCopied) {
+      THROW_ARANGO_EXCEPTION(TRI_ERROR_INPUT_REGISTERS_NOT_COPIED);
+    }
+    ++_baseIndex;
+    _inputRowCopied = false;
+    _numValuesWritten = 0;
+  }
+
+  // returns true if row was produced
+  bool produced() const {
+    return allValuesWritten() && _inputRowCopied;
+  }
+
+  std::unique_ptr<AqlItemBlock> stealBlock() {
```

Comment:
Please at comment about guarantees:

e.g. a returned block is always "full" if there are 100 lines it has a size of 100.
If there is no row the result will be NULLPTR